### PR TITLE
STORM-2501:  Auto populate Hive Credentials

### PR DESCRIPTION
**Pull request by omkreddy**

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
@omkreddy 
It might be better to file a new issue and change the title accordingly, since I'll mark STORM-2482 as resolved.

**Comment by omkreddy** [comment]
@HeartSaVioR @harshach @arunmahadevan  Pl review the PR.   

AutoHive Code contacts with Hive MetaStore to create the tokens. These tokens will be passed to workers. Since Hive depends on HDFS, We also need to configure Hadoop delegations tokens.

File: external/storm-hive/README.md [hunk 1]
```diff
@@ -101,6 +101,80 @@ Hive Trident state also follows similar pattern to HiveBolt it takes in HiveOpti
...
  ```
    
+   
+      
+##Working with Secure Hive
+If your topology is going to interact with secure Hive, your bolts/states needs to be authenticated by Hive Server. We 
+currently have 2 options to support this:
+
+### Using keytabs on all worker hosts
+If you have distributed the keytab files for hive user on all potential worker hosts then you can use this method. You should specify a
```
Comment (arunmahadevan) [inline external/storm-hive/README.md]: specify a -> specify

File: external/storm-hive/README.md [hunk 2]
```diff
@@ -101,6 +101,80 @@ Hive Trident state also follows similar pattern to HiveBolt it takes in HiveOpti
...
+
+If nimbus did not have the above configuration you need to add and then restart it. Ensure the hadoop configuration 
+files (core-site.xml, hdfs-site.xml and hive-site.xml) and the storm-hive connector jar with all the dependencies is present in nimbus's classpath.
+
+As an alternative to adding the configuration files (core-site.xml, hdfs-site.xml and hive-site.xml) to the classpath, you could specify the configurations
+as a part of the topology configuration. E.g. in you custom storm.yaml (or -c option while submitting the topology),
+
+```
+hiveCredentialsConfigKeys : ["cluster1", "cluster2"] (the hive clusters you want to fetch the tokens from)
+cluster1: [{"config1": "value1", "config2": "value2", ... }] (A map of config key-values specific to cluster1)
```
Comment (arunmahadevan) [inline external/storm-hive/README.md]: cluster value should be a map. Take a look at hdfs, hbase docs which was fixed recently.

File: external/storm-hive/README.md [hunk 3]
```diff
@@ -101,6 +101,80 @@ Hive Trident state also follows similar pattern to HiveBolt it takes in HiveOpti
...
+cluster2: [{"config1": "value1", "hive.keytab.file": "/path/to/keytab/for/cluster2/on/nimubs", "hive.kerberos.principal": "cluster2user@EXAMPLE.com", "hive.metastore.uris": "thrift://server:9083"}] (here along with other configs, we have custom keytab and principal for "cluster2" which will override the keytab/principal specified at topology level)
```
Comment (arunmahadevan) [inline external/storm-hive/README.md]: cluster value should be a map. Take a look at hdfs, hbase docs which was fixed recently.

File: external/storm-hive/README.md [hunk 4]
```diff
@@ -101,6 +101,80 @@ Hive Trident state also follows similar pattern to HiveBolt it takes in HiveOpti
...
+
+hdfsCredentialsConfigKeys : ["cluster1", "cluster2"] (the hdfs clusters you want to fetch the tokens from)
+cluster1: [{"config1": "value1", "config2": "value2", ... }] (A map of config key-values specific to cluster1)
```
Comment (arunmahadevan) [inline external/storm-hive/README.md]: cluster value should be a map. Take a look at hdfs, hbase docs which was fixed recently.

File: external/storm-hive/README.md [hunk 5]
```diff
@@ -101,6 +101,80 @@ Hive Trident state also follows similar pattern to HiveBolt it takes in HiveOpti
...
+cluster2: [{"config1": "value1", "hdfs.keytab.file": "/path/to/keytab/for/cluster2/on/nimubs", "hdfs.kerberos.principal": "cluster2user@EXAMPLE.com"}] (here along with other configs, we have custom keytab and principal for "cluster2" which will override the keytab/principal specified at topology level)
```
Comment (arunmahadevan) [inline external/storm-hive/README.md]: cluster value should be a map. Take a look at hdfs, hbase docs which was fixed recently.

**Comment by omkreddy** [comment]
@arunmahadevan Thanks for the review. Updated the PR.

File: external/storm-hive/README.md [hunk 6]
```diff
@@ -99,8 +99,85 @@ Hive Trident state also follows similar pattern to HiveBolt it takes in HiveOpti
...
+As an alternative to adding the configuration files (core-site.xml, hdfs-site.xml and hive-site.xml) to the classpath, you could specify the configurations
+as a part of the topology configuration. E.g. in you custom storm.yaml (or -c option while submitting the topology),
+
+
+```
+hiveCredentialsConfigKeys : ["cluster1", "cluster2"] (the hive clusters you want to fetch the tokens from)
+"cluster1": {"config1": "value1", "config2": "value2", ... } (A map of config key-values specific to cluster1)
+"cluster2": {"config1": "value1", "hive.keytab.file": "/path/to/keytab/for/cluster2/on/nimubs", "hive.kerberos.principal": "cluster2user@EXAMPLE.com", "hive.metastore.uris": "thrift://server:9083"} (here along with other configs, we have custom keytab and principal for "cluster2" which will override the keytab/principal specified at topology level)
+
+hdfsCredentialsConfigKeys : ["cluster1", "cluster2"] (the hdfs clusters you want to fetch the tokens from)
```
Comment (HeartSaVioR) [inline external/storm-hive/README.md]: @omkreddy 
Do we need to set cluster name same as hiveCredentialsConfigKeys? 
If then looks like cluster key-values map should have both hadoop config KVs and hive config KVs.
If it doesn't, let's use diffrent name.

Comment (omkreddy) [inline external/storm-hive/README.md]: @HeartSaVioR We can set different cluster names. Updated the docs

**Comment by HeartSaVioR** [comment]
+1 LGTM

**Pull request merged**